### PR TITLE
Adds prometheus export point to the agent pod.

### PR DESCRIPTION
**Pull request by meowjesty**
- Issue: #2975 

(wip, missing docs)

## Note for reviewer

I've tried keeping the changes in each system to each commit, hopefully that helps when reviewing this, since it touches a bunch of files.

### CODE REVIEW

**Comment by aviramha** [comment]
Note - this should be behind a flag.

File: mirrord/kube/src/api/container/job.rs [hunk 1]
```diff
@@ -152,6 +152,8 @@ where
                 "disabled".to_string(),
             ),
             ("app".to_string(), "mirrord".to_string()),
+            ("prometheus.io/scrape".to_string(), "true".to_string()),
+            ("prometheus.io/port".to_string(), "9000".to_string()),
```
Comment (Razz4780) [inline mirrord/kube/src/api/container/job.rs:156]: These labels may vary between scrapers, no? I think we should let the user set them via agent config (just mention port number and endpoint path in the doc)

File: mirrord/kube/src/api/container/job.rs [hunk 2]
```diff
@@ -163,6 +165,9 @@ where
         annotations.extend(BTreeMap::from([
             ("sidecar.istio.io/inject".to_string(), "false".to_string()),
             ("linkerd.io/inject".to_string(), "disabled".to_string()),
+            ("prometheus.io/scrape".to_string(), "true".to_string()),
+            // ("prometheus.io/path".to_string(), "/metrics".to_string()),
+            ("prometheus.io/port".to_string(), "9000".to_string()),
```
Comment (Razz4780) [inline mirrord/kube/src/api/container/job.rs:170]: Same, I think we should let the user set these

File: mirrord/kube/src/api/container/pod.rs [hunk 3]
```diff
@@ -106,6 +106,8 @@ impl ContainerVariant for PodVariant<'_> {
                     [
                         ("sidecar.istio.io/inject".to_string(), "false".to_string()),
                         ("linkerd.io/inject".to_string(), "disabled".to_string()),
+                        ("prometheus.io/scrape".to_string(), "true".to_string()),
+                        ("prometheus.io/port".to_string(), "9000".to_string()),
```
Comment (Razz4780) [inline mirrord/kube/src/api/container/pod.rs:110]: Same, I think we should let the user set these

File: mirrord/kube/src/api/container/pod.rs [hunk 4]
```diff
@@ -116,6 +118,8 @@ impl ContainerVariant for PodVariant<'_> {
                             "disabled".to_string(),
                         ),
                         ("app".to_string(), "mirrord".to_string()),
+                        ("prometheus.io/scrape".to_string(), "true".to_string()),
+                        ("prometheus.io/port".to_string(), "9000".to_string()),
```
Comment (Razz4780) [inline mirrord/kube/src/api/container/pod.rs:122]: Same, I think we should let the user set these

**Review by Razz4780 — CHANGES_REQUESTED** [review]
I didn't go over the code in agent submodules yet, because I have an important request that might require changes there - please no new actor framework ;_;
You're already using `axum` router, so why not go with [`State`](https://docs.rs/axum/latest/axum/extract/struct.State.html)? :eyes:
I'd just put all metrics in there (`AtomicU64`), wrap them together in `Arc` and share between the modules and the metrics endpoint

File: mirrord/agent/Cargo.toml [hunk 5]
```diff
@@ -70,6 +70,10 @@ x509-parser = "0.16"
 rustls.workspace = true
 envy = "0.4"
 socket2.workspace = true
+prometheus = { version = "0.13", features = ["process"] }
+kameo = { git = "https://github.com/tqwewe/kameo", branch = "main" }
```
[the remark this answers is not in the document]
Comment (meowjesty) [reply mirrord/agent/Cargo.toml:74]: Removed.

File: mirrord/agent/README.md [hunk 6]
```diff
@@ -6,3 +6,223 @@ Agent part of [mirrord](https://github.com/metalbear-co/mirrord) responsible for
 mirrord-agent is written in Rust for safety, low memory consumption and performance.
 
 mirrord-agent is distributed as a container image (currently only x86) that is published on [GitHub Packages publicly](https://github.com/metalbear-co/mirrord-agent/pkgs/container/mirrord-agent). 
+
+## Enabling prometheus metrics
+
+TODO(alex) [mid]: Talk how to enable it from env whatever.
```
[the remark this answers is not in the document]
Comment (meowjesty) [reply mirrord/agent/README.md]: Removed

File: mirrord/kube/src/api/container/job.rs (hunk 1 above)
Comment (meowjesty) [reply mirrord/kube/src/api/container/job.rs]: Yeah, was there for convenience while I was trying out stuff, removed in favour of config.

File: mirrord/agent/src/entrypoint.rs [hunk 7]
```diff
@@ -488,10 +489,19 @@ impl ClientConnectionHandler {
...
 
 /// Initializes the agent's [`State`], channels, threads, and runs [`ClientConnectionHandler`]s.
-#[tracing::instrument(level = "trace", ret)]
-async fn start_agent(args: Args) -> Result<()> {
+#[tracing::instrument(level = Level::TRACE, ret, err)]
+async fn start_agent(args: Args) -> AgentResult<()> {
     trace!("start_agent -> Starting agent with args: {args:?}");
 
+    if let Some(metrics_address) = args.metrics.as_ref() {
+        let address = metrics_address.parse()?;
```
Comment (Razz4780) [inline mirrord/agent/src/entrypoint.rs:497]: You can store the address as `SocketAddr` in both agent config and agent CLI params, no need to parse here and add a very generic `AgentError` variant :eyes:

File: mirrord/agent/src/entrypoint.rs [hunk 8]
```diff
@@ -488,10 +489,19 @@ impl ClientConnectionHandler {
...
+        tokio::spawn(async move {
+            start_metrics(address)
+                .await
+                .inspect_err(|fail| tracing::error!(?fail, "Failed starting metrics server!"))
+        });
```
Comment (Razz4780) [inline mirrord/agent/src/entrypoint.rs:502]: Something should watch over the metrics task. Maybe the `CancellationToken` we have in `start_agent`?

File: mirrord/agent/src/file.rs [hunk 9]
```diff
@@ -147,24 +135,27 @@ pub fn resolve_path<P: AsRef<Path> + std::fmt::Debug, R: AsRef<Path> + std::fmt:
 
 impl FileManager {
     /// Executes the request and returns the response.
-    #[tracing::instrument(level = "trace", skip(self))]
-    pub fn handle_message(&mut self, request: FileRequest) -> Result<Option<FileResponse>> {
+    #[tracing::instrument(level = Level::TRACE, skip(self), err)]
+    pub(crate) async fn handle_message(
```
Comment (Razz4780) [inline mirrord/agent/src/file.rs:139]: This and other `FileManager` functions are now unnecessarily async

File: mirrord/agent/src/metrics.rs [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+#[tracing::instrument(level = Level::TRACE, skip_all, ret ,err)]
+pub(crate) async fn start_metrics(address: SocketAddr) -> Result<(), axum::BoxError> {
+    let app = Router::new().route("/metrics", get(get_metrics));
+
+    let listener = TcpListener::bind(address)
+        .await
+        .map_err(AgentError::from)
+        .inspect_err(|fail| tracing::error!(?fail, "Actor listener!"))?;
+
+    tokio::spawn(async move {
```
Comment (Razz4780) [inline mirrord/agent/src/metrics.rs:115]: This function already runs in `tokio::spawn`

File: mirrord/agent/src/metrics.rs [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+#[derive(Error, Debug)]
+pub(crate) enum MetricsError {
+    #[error(transparent)]
+    FromUtf8(#[from] std::string::FromUtf8Error),
+
+    #[error(transparent)]
+    Prometheus(#[from] prometheus::Error),
+}
+
+unsafe impl Send for MetricsError {}
```
Comment (Razz4780) [inline mirrord/agent/src/metrics.rs:84]: wut :eyes:

File: mirrord/agent/src/sniffer.rs [hunk 12]
```diff
@@ -276,6 +277,7 @@ where
             } => {
                 if self.port_subscriptions.subscribe(client_id, port) {
                     self.update_packet_filter()?;
+                    MIRROR_CONNECTION_SUBSCRIPTION.inc();
```
Comment (Razz4780) [inline mirrord/agent/src/sniffer.rs:280]: Different metric, should be port subscription

File: mirrord/agent/src/sniffer/api.rs [hunk 13]
```diff
@@ -15,11 +15,15 @@ use tokio_stream::{
...
+use crate::{
+    error::AgentError, metrics::MIRROR_PORT_SUBSCRIPTION, util::ClientId, watched_task::TaskStatus,
+};
 
 /// Interface used by clients to interact with the
 /// [`TcpConnectionSniffer`](super::TcpConnectionSniffer). Multiple instances of this struct operate
 /// on a single sniffer instance.
+///
+/// Enabled by the `mirror` feature for incoming traffic.
 pub(crate) struct TcpSnifferApi {
```
Comment (Razz4780) [inline mirrord/agent/src/sniffer/api.rs:27]: I'd move all metrics updated to the sniffer itself

File: mirrord/agent/src/steal/http/filter.rs [hunk 14]
```diff
@@ -3,7 +3,7 @@ use hyper::Request;
 use tracing::Level;
 
 /// Currently supported filtering criterias.
-#[derive(Debug)]
+#[derive(Debug, Clone)]
```
Comment (Razz4780) [inline mirrord/agent/src/steal/http/filter.rs:6]: :eyes:  why do you need this?

File: mirrord/agent/src/steal/connection.rs [hunk 15]
```diff
@@ -649,14 +683,26 @@ impl TcpConnectionStealer {
...
 
             Command::PortUnsubscribe(port) => {
-                self.port_subscriptions.remove(client_id, port).await?;
+                if let Some(filtered) = self.port_subscriptions.remove(client_id, port).await? {
+                    if filtered {
+                        STEAL_FILTERED_PORT_SUBSCRIPTION.inc();
+                    } else {
+                        STEAL_UNFILTERED_PORT_SUBSCRIPTION.inc();
+                    }
+                }
```
Comment (Razz4780) [inline mirrord/agent/src/steal/connection.rs:705]: I'd move updating port subscription metrics to `PortSubscriptions`. You won't have to return the bools from its methods

File: mirrord/agent/src/metrics.rs [hunk 16]
```diff
@@ -0,0 +1,123 @@
...
+use thiserror::Error;
+use tokio::net::TcpListener;
+use tracing::Level;
+
+use crate::error::AgentError;
+
+pub(crate) static OPEN_FD_COUNT: LazyLock<IntGauge> = LazyLock::new(|| {
+    register_int_gauge!(
+        "mirrord_agent_open_fd_count",
+        "amount of open fds in mirrord-agent"
```
Comment (Razz4780) [inline mirrord/agent/src/metrics.rs:14]:
```suggestion
        "amount of open file descriptors in mirrord-agent file manager"
```

File: mirrord/agent/src/metrics.rs [hunk 17]
```diff
@@ -0,0 +1,123 @@
...
+        "mirrord_agent_mirror_port_subscription_count",
+        "amount of mirror port subscriptions in mirror-agent"
+    )
+    .expect("Valid at initialization")
+});
+
+pub(crate) static MIRROR_CONNECTION_SUBSCRIPTION: LazyLock<IntGauge> = LazyLock::new(|| {
+    register_int_gauge!(
+        "mirrord_agent_mirror_connection_subscription_count",
+        "amount of connections in steal mode in mirrord-agent"
```
Comment (Razz4780) [inline mirrord/agent/src/metrics.rs:30]:
```suggestion
        "amount of connections in mirror mode in mirrord-agent"
```

File: mirrord/agent/src/metrics.rs [hunk 18]
```diff
@@ -0,0 +1,123 @@
...
+
+pub(crate) static STEAL_UNFILTERED_PORT_SUBSCRIPTION: LazyLock<IntGauge> = LazyLock::new(|| {
+    register_int_gauge!(
+        "mirrord_agent_steal_unfiltered_port_subscription_count",
+        "amount of unfiltered steal port subscriptions in mirrord-agent"
+    )
+    .expect("Valid at initialization!")
+});
+
+pub(crate) static STEAL_CONNECTION_SUBSCRIPTION: LazyLock<IntGauge> = LazyLock::new(|| {
```
Comment (Razz4780) [inline mirrord/agent/src/metrics.rs:51]: Should be split into filtered and unfiltered

**Review by Razz4780 — CHANGES_REQUESTED** [review]
I think we're missing:
1. Stolen HTTP requests in progress
2. Connected clients
3. DNS lookups in progress

I see you went for global gauges, that's fine as well

**Review by Razz4780 — COMMENTED** [review]
If you don't like statics, you can go with sth like this:
```rust
#[derive(Clone)]
struct Metrics {
    open_file_descriptors: IntCounter,
    // other counters
}

// pass registry to the metrics http server
// pass metrics clones to agent tasks
fn setup_metrics_registry() -> (Registry, Metrics) {
    let open_file_descriptors = IntCounter::new("name", "help").unwrap();

    let registry = Registry::new();
    registry.register(Box::new(open_file_descriptors.clone())).unwrap();

    let metrics = Metrics {
        open_file_descriptors,
    };

    (registry, metrics)
}

// metrics endpoint handler
async fn get_metrics(State(registry): State<Registry>) -> (StatusCode, String) {
    let mut buffer = String::new();
    let encoder = TextEncoder::new();
    let metric_families = registry.gather();
    
    match encoder.encode_utf8(&metric_families, &mut buffer) {
        Ok(()) => (StatusCode::OK, buffer),
        Err(error) => {
            (StatusCode::INTERNAL_SERVER_ERROR, error.to_string())
        }
    }
}

// metrics http server router
async fn metrics_router(registry: Registry) -> Router {
    Router::new().route("/metrics", get(get_metrics)).with_state(registry)
}
```
Imho looks nicer and will allow for unit testing (statics would be shared across tests)

File: mirrord/agent/src/entrypoint.rs (hunk 8 above)
Comment (meowjesty) [reply mirrord/agent/src/entrypoint.rs]: Help, how do I make this happen? Just `clone` the cancellation token before `tokio::spawn`?

Comment (Razz4780) [reply mirrord/agent/src/entrypoint.rs]: Yep

Comment (meowjesty) [reply mirrord/agent/src/entrypoint.rs]: Added the `cancellation_token` to axum.

File: mirrord/agent/src/file.rs [hunk 19]
```diff
@@ -261,18 +246,26 @@ impl FileManager {
...
-            ..Default::default()
+            open_files: Default::default(),
+            dir_streams: Default::default(),
+            getdents_streams: Default::default(),
+            fds_iter: (0..=u64::MAX),
         }
     }
 
-    #[tracing::instrument(level = "trace", skip(self))]
+    // TODO(alex) [mid]: Fails with the wrong error?
```
Comment (Razz4780) [inline mirrord/agent/src/file.rs:263]: TODO :eyes:

File: mirrord/agent/src/metrics.rs [hunk 20]
```diff
@@ -0,0 +1,177 @@
...
+    use prometheus::{Encoder, TextEncoder};
+
+    let metric_families = prometheus::gather();
+
+    let mut buffer = Vec::new();
+    TextEncoder
+        .encode(&metric_families, &mut buffer)
+        .inspect_err(|error| tracing::error!(%error, "unable to encode prometheus metrics"))?;
+
+    Ok(String::from_utf8(buffer)?)
```
Comment (Razz4780) [inline mirrord/agent/src/metrics.rs:112]: You can use [`TextEncode::encode_to_string`](https://docs.rs/prometheus/latest/prometheus/struct.TextEncoder.html#method.encode_to_string) (you won't need `MetricsError` anymore)

File: mirrord/agent/src/metrics.rs [hunk 21]
```diff
@@ -0,0 +1,177 @@
...
+        .await
+        .map_err(AgentError::from)
+        .inspect_err(|fail| tracing::error!(?fail, "Actor listener!"))?;
+
+    axum::serve(listener, app)
+        .with_graceful_shutdown(async move { cancellation_token.cancelled().await })
+        .await
+        .inspect_err(|fail| {
+            tracing::error!(%fail, "Could not start agent metrics
+        server!")
```
Comment (Razz4780) [inline mirrord/agent/src/metrics.rs:132]: I think you should cancel the token on error as well. Otherwise we'll have an agent running without metrics

File: mirrord/agent/src/metrics.rs [hunk 22]
```diff
@@ -0,0 +1,177 @@
...
+pub(crate) async fn start_metrics(
+    address: SocketAddr,
+    cancellation_token: CancellationToken,
+) -> Result<(), axum::BoxError> {
+    let app = Router::new().route("/metrics", get(get_metrics));
+
+    let listener = TcpListener::bind(address)
+        .await
+        .map_err(AgentError::from)
+        .inspect_err(|fail| tracing::error!(?fail, "Actor listener!"))?;
```
Comment (Razz4780) [inline mirrord/agent/src/metrics.rs:125]:
```suggestion
        .inspect_err(|fail| tracing::error!(?fail, "Failed to bind TCP socket for metrics server"))?;
```

File: mirrord/agent/src/sniffer.rs [hunk 23]
```diff
@@ -287,6 +292,8 @@ where
             } => {
                 if self.port_subscriptions.unsubscribe(client_id, port) {
                     self.update_packet_filter()?;
+
+                    MIRROR_PORT_SUBSCRIPTION.dec();
```
Comment (Razz4780) [inline mirrord/agent/src/sniffer.rs:296]: Now that I think of it, looks like you also need to decrement in `handle_client_closed`

File: mirrord/agent/src/sniffer.rs [hunk 24]
```diff
@@ -268,6 +269,8 @@ where
                 command: SnifferCommandInner::NewClient(sender),
             } => {
                 self.handle_new_client(client_id, sender);
+
+                MIRROR_CONNECTION_SUBSCRIPTION.inc();
```
Comment (Razz4780) [inline mirrord/agent/src/sniffer.rs:273]: I don't see corresponding `.dec()` anywhere

File: mirrord/agent/src/steal/subscriptions.rs [hunk 25]
```diff
@@ -215,6 +219,11 @@ impl<R: PortRedirector> PortSubscriptions<R> {
     /// * `client_id` - identifier of the client that issued the subscription
     /// * `port` - number of the subscription port
     ///
+    /// # Returns
+    ///
+    /// `Some(true)` if the subscprition has an HTTP filter, `Some(false)` if it's unfiltered, and
+    /// `None` if we could not find the [`PortSubscription`].
+    ///
```
Comment (Razz4780) [inline mirrord/agent/src/steal/subscriptions.rs:226]: Leftover

File: mirrord/agent/src/steal/connections.rs [hunk 26]
```diff
@@ -304,9 +312,21 @@ impl StolenConnections {
...
+
             match task.run().await {
                 Ok(()) => {
                     tracing::trace!(connection_id, "Connection task finished");
+
+                    if filtered {
+                        STEAL_FILTERED_CONNECTION_SUBSCRIPTION.dec();
+                    } else {
+                        STEAL_UNFILTERED_CONNECTION_SUBSCRIPTION.dec();
+                    }
```
Comment (Razz4780) [inline mirrord/agent/src/steal/connections.rs:329]: I think you need to handle these counters elsewhere (maybe `ConnectionTask`?). Even if we have a filtered port subscription, the connection might still be rejected as not HTTP (and proxied raw to the original target). See `ConnectionTask::run` code

**Review by Razz4780 — CHANGES_REQUESTED** [review]

File: mirrord/agent/src/metrics.rs (hunk 21 above)
Comment (Razz4780) [reply mirrord/agent/src/metrics.rs:132]: In general, a failure in this function should terminate the agent (and cancelling the token should terminate function)

File: mirrord/agent/src/metrics.rs (hunk 20 above)
Comment (meowjesty) [reply mirrord/agent/src/metrics.rs:112]: It also returns a `prometheus::error`, how can I avoid `MetricsError` with it?

Comment (Razz4780) [reply mirrord/agent/src/metrics.rs:112]: I'd just change the return type to `(StatusCode, String)`. The function could be very small, just
```rust
let metric_families = prometheus::gather();
match TextEncoder.encode_to_string(&metric_families) {
    Ok(response) => (StatusCode::OK, response),
    Err(error) => {
        tracing::error!(...);
        (StatusCode::INTERNAL_SERVER_ERROR, error.to_string())
    }
}
```

File: mirrord/agent/src/sniffer.rs (hunk 24 above)
Comment (meowjesty) [reply mirrord/agent/src/sniffer.rs:273]: Added to `handle_client_closed`.

File: mirrord/agent/src/steal/connections.rs (hunk 26 above)
Comment (meowjesty) [reply mirrord/agent/src/steal/connections.rs]: Think it's in the right place now?

File: mirrord/agent/src/file.rs (hunk 19 above)
Comment (meowjesty) [reply mirrord/agent/src/file.rs]: Opened an issue for this.

File: mirrord/agent/src/metrics.rs [hunk 27]
```diff
@@ -0,0 +1,166 @@
...
+pub(crate) static STEAL_FILTERED_CONNECTION_SUBSCRIPTION: LazyLock<IntGauge> =
+    LazyLock::new(|| {
+        register_int_gauge!(
+            "mirrord_agent_steal_connection_subscription_count",
+            "amount of filtered connections in steal mode in mirrord-agent"
+        )
+        .expect("Valid at initialization!")
+    });
+
+pub(crate) static STEAL_UNFILTERED_CONNECTION_SUBSCRIPTION: LazyLock<IntGauge> =
```
Comment (Razz4780) [inline mirrord/agent/src/metrics.rs:61]: The way you manipulate this metric is hard to follow and looks very prone to errors :V
There's a lot of branching and early exits in the code.

I suggest you just:
1. Add `UnfilteredStealTask::new`
2. Inc the gauge there
3. Dec the gauge in `impl Drop for UnfilteredStealTask`

File: mirrord/agent/src/metrics.rs [hunk 28]
```diff
@@ -0,0 +1,166 @@
...
+
+pub(crate) static STEAL_UNFILTERED_PORT_SUBSCRIPTION: LazyLock<IntGauge> = LazyLock::new(|| {
+    register_int_gauge!(
+        "mirrord_agent_steal_unfiltered_port_subscription_count",
+        "amount of unfiltered steal port subscriptions in mirrord-agent"
+    )
+    .expect("Valid at initialization!")
+});
+
+pub(crate) static STEAL_FILTERED_CONNECTION_SUBSCRIPTION: LazyLock<IntGauge> =
```
Comment (Razz4780) [inline mirrord/agent/src/metrics.rs:52]: The way you manipulate this metric is hard to follow and looks very prone to errors :V
There's a lot of branching and early exits in the code.

I suggest you just:
1. Inc the gauge in `FilteredStealTask::new`
2. In `FilteredStealTask::run` after `let res = self.run_until_http_ends(...).await;`, dec the gauge and mark that decrement was done (new struct field)
3. If decrement was not done yet, dec the gauge in `impl Drop for FilteredStealTask`

**Review by Razz4780 — CHANGES_REQUESTED** [review]

File: mirrord/agent/src/metrics.rs [hunk 29]
```diff
@@ -0,0 +1,166 @@
...
+    .expect("Valid at initialization!")
+});
+
+pub(crate) static UDP_OUTGOING_CONNECTION: LazyLock<IntGauge> = LazyLock::new(|| {
+    register_int_gauge!(
+        "mirrord_agent_udp_outgoing_connection_count",
+        "amount of udp outgoing connections in mirrord-agent"
+    )
+    .expect("Valid at initialization!")
+});
```
[the remark this answers is not in the document]
Comment (meowjesty) [reply mirrord/agent/src/metrics.rs:108]: Alright, think I've added these now.

File: mirrord/agent/src/metrics.rs [hunk 30]
```diff
@@ -0,0 +1,189 @@
...
+pub(crate) static STEAL_UNFILTERED_CONNECTION_SUBSCRIPTION: LazyLock<IntGauge> =
+    LazyLock::new(|| {
+        register_int_gauge!(
+            "mirrord_agent_steal_connection_subscription_count",
+            "amount of unfiltered connections in steal mode in mirrord-agent"
+        )
+        .expect("Valid at initialization!")
+    });
+
+pub(crate) static HTTP_REQUEST_IN_PROGRESS_COUNT: LazyLock<IntGauge> = LazyLock::new(|| {
```
Comment (Razz4780) [inline mirrord/agent/src/metrics.rs:86]: This metric is decremented in a wrong place. I tried to find better places to update it, but oh man. What a mess this is is :sweat_smile: 

Let's just count streaming requests, since this should cover all requests in non-ancient mirrord versions anyway. This can be done easily, because the value of the gauge should always be equal to aggregate size of all `TcpStealerApi::response_body_txs` - update gauge whenever we update the mapping, including `TcpStealerApi` drop

**Review by Razz4780 — CHANGES_REQUESTED** [review]

File: mirrord/agent/src/metrics.rs (hunk 30 above)
Comment (meowjesty) [reply mirrord/agent/src/metrics.rs:86]: Changed.

File: mirrord/agent/src/file.rs [hunk 31]
```diff
@@ -572,20 +572,30 @@ impl FileManager {
...
+    }
 
+    /// Always returns `None`, since we don't return any [`FileResponse`] back to mirrord
+    /// on `close_dir` of an fd.
+    #[tracing::instrument(level = Level::TRACE, skip(self))]
+    pub(crate) fn close_dir(&mut self, fd: u64) -> Option<FileResponse> {
         if self.dir_streams.remove(&fd).is_none() && self.getdents_streams.remove(&fd).is_none() {
             error!("FileManager::close_dir -> fd {:#?} not found", fd);
+        } else {
+            OPEN_FD_COUNT.dec();
```
[the remark this answers is not in the document]
Comment (meowjesty) [reply mirrord/agent/src/file.rs:597]: Changed.

File: mirrord/agent/src/metrics.rs [hunk 32]
```diff
@@ -0,0 +1,189 @@
...
+
+pub(crate) static STEAL_FILTERED_PORT_SUBSCRIPTION: LazyLock<IntGauge> = LazyLock::new(|| {
+    register_int_gauge!(
+        "mirrord_agent_steal_filtered_port_subscription_count",
+        "amount of filtered steal port subscriptions in mirrord-agent"
+    )
+    .expect("Valid at initialization!")
+});
+
+pub(crate) static STEAL_UNFILTERED_PORT_SUBSCRIPTION: LazyLock<IntGauge> = LazyLock::new(|| {
```
[the remark this answers is not in the document]
Comment (meowjesty) [reply mirrord/agent/src/metrics.rs:60]: Did I just add something that we had removed back again?

File: mirrord/agent/src/metrics.rs [hunk 33]
```diff
@@ -0,0 +1,189 @@
...
+
+pub(crate) static MIRROR_CONNECTION_SUBSCRIPTION: LazyLock<IntGauge> = LazyLock::new(|| {
+    register_int_gauge!(
+        "mirrord_agent_mirror_connection_subscription_count",
+        "amount of connections in mirror mode in mirrord-agent"
+    )
+    .expect("Valid at initialization!")
+});
+
+pub(crate) static STEAL_FILTERED_PORT_SUBSCRIPTION: LazyLock<IntGauge> = LazyLock::new(|| {
```
[the remark this answers is not in the document]
Comment (meowjesty) [reply mirrord/agent/src/metrics.rs:52]: Same here.

File: mirrord/agent/src/metrics.rs [hunk 34]
```diff
@@ -0,0 +1,189 @@
...
+
+pub(crate) static TCP_OUTGOING_CONNECTION: LazyLock<IntGauge> = LazyLock::new(|| {
+    register_int_gauge!(
+        "mirrord_agent_tcp_outgoing_connection_count",
+        "amount of tcp outgoing connections in mirrord-agent"
+    )
+    .expect("Valid at initialization!")
+});
+
+pub(crate) static UDP_OUTGOING_CONNECTION: LazyLock<IntGauge> = LazyLock::new(|| {
```
[the remark this answers is not in the document]
Comment (meowjesty) [reply mirrord/agent/src/metrics.rs:111]: Changed.

**Review by Razz4780 — APPROVED** [review]
Let this saga end

File: mirrord/agent/src/metrics.rs [hunk 35]
```diff
@@ -0,0 +1,276 @@
...
+    steal_unfiltered_port_subscription: IntGauge,
+    steal_filtered_connection_subscription: IntGauge,
+    steal_unfiltered_connection_subscription: IntGauge,
+    http_request_in_progress_count: IntGauge,
+    tcp_outgoing_connection: IntGauge,
+    udp_outgoing_connection: IntGauge,
+}
+
+impl Metrics {
+    fn new() -> Self {
```
Comment (Razz4780) [inline mirrord/agent/src/metrics.rs:65]: If we keep this metrics in a struct, I'd rather not use the default `prometheus` registry (this is what happens with `register_int_gauge!`.
1. Add [`Registry`](https://docs.rs/prometheus/latest/prometheus/struct.Registry.html) as a field in `Metrics`
2. Register all gauges there

**Review by Razz4780 — CHANGES_REQUESTED** [review]
One nit for the new changes

File: mirrord/agent/src/metrics.rs (hunk 35 above)
Comment (meowjesty) [reply mirrord/agent/src/metrics.rs:76]: Changed, now we even skip over `AlreadyReg` error which should make it extra nice for tests, and not affect real-world usage.

**Review by Razz4780 — APPROVED** [review]
The saga ends
